### PR TITLE
SI-6493 Fix existential type used to hide local classes

### PR DESCRIPTION
**Pull request by retronym**
In:

```
def foo = { object O { class C }; new O.C }
```

The inferred return type of can't refer to the local
symbols; they must not leak out of the compilation unit.

Instead, `packSymbols` is used to choose a less precise
existential type, `(AnyRef { type C <: AnyRef })#C)`.

This is implemented as a `TypeMap`, which is supposed to
takes care of rebinding `C` to something valid in the new
prefix (`(AnyRef { type C <: AnyRef })`).

If `C` was orginally a type alias, and the original
prefix was a refinement type, this was handled in
`AliasTypeRef#coevolveSym`, which looks for a type
in the new prefix with the name `C`. But for other
type refs (e.g. a class type ref, as in this case),
a no-op `coevolveSym` was used, deferring the rebinding
of abstract classes until `typeRef`.

But our case falls between the cracks, and we end up
propagating a type ref in which the prefix does not
contain the member.

With the help of `-uniqid`, this is clear:

```
<method> def foo#7445(): scala#21.this.AnyRef#2222{type Bar#12153 <: scala#21.this.AnyRef#2222}#Bar#12125
```

Notice the reference to symbol `#12125`, rather than `#12153`.

This commit moves the `coevolveSym` logic up to `TypeRef`,
generalizing it to work for any `pre1`. This example answered
the question in that code:

> // TODO: is there another way a typeref's symbol can refer to a symbol defined in its pre?

Review by @adriaanm. Found this patch in a dusty corner.

### CODE REVIEW

**Comment by retronym** [comment]
Interesting failure in some interesting code.

**Comment by retronym** [comment]
Minimized:

``` scala
trait T[T] {
  trait U {
    def x = ()
  }
  def u: U
}

object Test {
  def foo = new T[Any] {
    def u: U = ???
  }

  // okay
  (foo: T[Any]).u.x

  // fail
  foo.u.x
}
```

**Comment by retronym** [comment]
Let's have another spin.

File: test/files/run/t6493/A_1.scala [hunk 1]
```diff
@@ -0,0 +1,3 @@
+object one {
+  def foo() = { object Foo { class Bar } ; new Foo.Bar }
```
Comment (xeno-by) [inline test/files/run/t6493/A_1.scala:2]: I think it might be nice to turn this into a REPL test, so that we can see when the inference behaviour changes.

**Comment by xeno-by** [comment]
Btw I have a somewhat related problem in https://issues.scala-lang.org/browse/SI-7736 with a vampire structural type. The `@macroImpl` annotation blindly copied from the original macro def refers to wrong type parameter symbols, which prevents the macro engine from passing correct type arguments to the corresponding macro impl.

**Comment by xeno-by** [comment]
Just went to see how `packedType` and friends work to make sure that annotations are indeed copied to structural type members. Indeed they are since it all boils down to `cloneSymbol`: https://github.com/xeno-by/scala/blob/9f62900145029786457f27c1fe3083cde5d83fe2/src/reflect/scala/reflect/internal/Symbols.scala#L1693.

**Comment by retronym** [comment]
Next failure:

``` scala
trait U {
  val global: SymTab
  object o { def apply(a: Any) = ()}
  object p
}

trait SymTab

trait T { self: SymTab =>
  def e: U { val global: T.this.type} = ???
  e.o(null)  // error: U{val global: <refinement>.type}#<none> does not take parameters
}

```

**Pull request closed without merging**